txnsync/peerscheduler_test.go

```diff
@@ -0,0 +1,194 @@
+// Copyright (C) 2019-2021 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package txnsync
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/algorand/go-algorand/test/partitiontest"
+)
+
+// TestBasics tests that the push, pop, len, swap and less functions perform appropriately
+func TestBasics(t *testing.T) {
+	partitiontest.PartitionTest(t)
+
+	ps := peerScheduler{}
+
+	require.Equal(t, 0, ps.Len())
+
+	peers := []Peer{
+		Peer{
+			lastSentMessageSequenceNumber: 123,
+		},
+		Peer{
+			lastSentMessageSequenceNumber: 456,
+		},
+	}
+
+	ps.Push(peerBucket{&peers[0], 0 * time.Millisecond})
+	ps.Push(peerBucket{&peers[1], 1 * time.Millisecond})
+
+	require.Equal(t, 2, ps.Len())
+
+	require.Equal(t, uint64(123), ps.peers[0].peer.lastSentMessageSequenceNumber)
+	require.Equal(t, uint64(456), ps.peers[1].peer.lastSentMessageSequenceNumber)
+	require.True(t, ps.Less(0, 1))
+
+	ps.Swap(0, 1)
+	require.Equal(t, uint64(123), ps.peers[1].peer.lastSentMessageSequenceNumber)
+	require.Equal(t, uint64(456), ps.peers[0].peer.lastSentMessageSequenceNumber)
+	require.True(t, ps.Less(1, 0))
+
+	backPeer := ps.Pop().(peerBucket)
+
+	require.Equal(t, uint64(123), backPeer.peer.lastSentMessageSequenceNumber)
+
+	backPeer = ps.Pop().(peerBucket)
+
+	require.Equal(t, uint64(456), backPeer.peer.lastSentMessageSequenceNumber)
+
+	require.Equal(t, 0, ps.Len())
+}
+
+// TestSchedulerBasics tests the basic scheduler helper functions
+func TestSchedulerBasics(t *testing.T) {
+	partitiontest.PartitionTest(t)
+
+	ps := peerScheduler{}
+
+	peers := []Peer{
+		Peer{
+			lastSentMessageSequenceNumber: 123,
+		},
+		Peer{
+			lastSentMessageSequenceNumber: 456,
+		},
+		Peer{
+			lastSentMessageSequenceNumber: 789,
+		},
+	}
+
```

Comment:
nit: some comments would be helpful.

i.e.
// check that the duration is recorded
// ensure the duration has been cleared

And below
// ensure duplicate peers are removed

etc.

Reply:
I'm going to merge in this PR, but I agree with @winder's suggestion  - we should have more comments describing the purpose of the test.

We can add these comments in a separate follow-up PR.